File: packages/mirascript/src/vm/operations/call.ts
import { VmError } from '../../helpers/error.js';
import { display } from '../../helpers/serialize.js';
import { isVmExtern, isVmConst } from '../../helpers/types.js';
import type { VmExtern, VmFunction, VmAny, VmArray, VmValue } from '../types/index.js';
import { $AssertInit } from './common.js';

/** 调用返回类型 */
type CallReturn<T extends VmValue> =
    T extends VmFunction<infer F>
        ? F extends (...args: readonly VmAny[]) => infer R
            ? // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
              R extends void | undefined
                ? null
                : R
            : VmValue
        : T extends VmExtern<infer E>
          ? E extends (...args: readonly VmAny[]) => infer R
              ? // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
                R extends void | undefined
                  ? null
                  : R extends VmValue
                    ? R
                    : R extends object
                      ? VmExtern<R>
                      : VmValue
              : VmValue
          : VmValue;

/** 调用函数 */
export function $Call<T extends VmValue, A extends readonly VmValue[]>(func: T, args: A): CallReturn<T> {
    const argsLen = args.length;
    for (let i = 0; i < argsLen; i++) {
        $AssertInit(args[i]);
    }
    if (typeof func == 'function') {
        return (func(...args) ?? null) as CallReturn<T>;
    }
    if (isVmExtern(func)) {
        return (func.call(args) ?? null) as CallReturn<T>;
    }
    throw new VmError(`Value is not callable: ${display(func)}`, null);
}

/** 过滤剩余参数数组 */
export function $VArgs(varags: VmAny[]): VmArray {
    for (let i = 0, l = varags.length; i < l; i++) {
        const el = varags[i];
        if (!isVmConst(el)) {
            varags[i] = null;
        }
    }
    return varags as VmArray;
}

File: crates/python/mirascript/vm/operations/call.py
from ...helpers.types import is_vm_const, is_vm_function
from ...helpers.constants import kVmFunction
from ..error import VmError
from ..types.types import VmAny, VmValue, VmArray
from .common import AssertInit


def Call(func: VmAny, *args: VmAny) -> VmValue:
    for a in args:
        AssertInit(a)
    if is_vm_function(func):
        info = getattr(func, kVmFunction)
        arg_count = len(args)
        if info.min_args <= arg_count <= info.max_args:
            return func(*args)
        if arg_count < info.min_args:
            args = args + (None,) * (info.min_args - arg_count)
            return func(*args)
        if arg_count > info.max_args:
            args = args[: info.max_args]
            return func(*args)
    raise VmError(f"{type(func)}, {func} object is not callable", None)


def Vargs(varags: "list[VmAny]") -> "VmArray":
    """将非常量元素置为 None，返回新列表"""
    return [var if is_vm_const(var) else None for var in varags]
